### PR TITLE
handling tool errors

### PR DESCRIPTION
**Pull request by jngz-es**
### Description
The agent catches all errors or exceptions from tools and sends the error messages to the LLM.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java [hunk 1]
```diff
@@ -470,27 +471,58 @@ private void runReAct(
...
-                                // TODO: support tool parameter override : langauge_model_tool.prompt
-                                llmToolTmpParameters.put(MLAgentExecutor.QUESTION, actionInput);
-                                tools.get(action).run(llmToolTmpParameters, nextStepListener); // run tool
-                            } else {
-                                tools.get(action).run(toolParams, nextStepListener); // run tool
+                            try {
+                                String action1 = action;
+                                ActionListener<Object> toolListener = ActionListener.wrap(r -> {
+                                    ((ActionListener<Object>) nextStepListener).onResponse(r);
+                                    ;
```
Comment (Hailong-am) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:478]: duplicate `;` here

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java [hunk 2]
```diff
@@ -470,27 +471,58 @@ private void runReAct(
...
+                            String res = String
+                                .format(Locale.ROOT, "Failed to run the tool %s due to wrong input %s.", action, actionInput);
+                            lastActionResult.set(res);
                             lastTool.set(action);
-                            String res = "Tool " + action + " can't work for input: " + actionInput;
                             ((ActionListener<Object>) nextStepListener).onResponse(res);
                         }
                     } else {
-                        lastTool.set(null);
+                        lastTool.set(action);
```
Comment (Hailong-am) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:521]: what does `lastTool` used for? seem it never used. same for `lastToolListener` and `lastActionResult`

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java (hunk 1 above)
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:478]: Oops, removed it.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java (hunk 2 above)
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:521]: Yeah, they are legacy code unused. We plan to refactor them afterwards. Anyway I removed them in this pr.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java [hunk 3]
```diff
@@ -470,28 +468,51 @@ private void runReAct(
...
-                                Map<String, String> llmToolTmpParameters = new HashMap<>();
-                                llmToolTmpParameters.putAll(tmpParameters);
-                                llmToolTmpParameters.putAll(toolSpecMap.get(action).getParameters());
-                                // TODO: support tool parameter override : langauge_model_tool.prompt
-                                llmToolTmpParameters.put(MLAgentExecutor.QUESTION, actionInput);
-                                tools.get(action).run(llmToolTmpParameters, nextStepListener); // run tool
-                            } else {
-                                tools.get(action).run(toolParams, nextStepListener); // run tool
+                            try {
+                                String action1 = action;
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java:472]: minor: make variable name more readable, `action1` -> `finalAction`

Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java]: done

**Review by Hailong-am — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Pull request merged**